### ISSUE TITLE
Optimize main_chain update of new blocks

### ISSUE THREAD

**Issue by tdroxler** [alephium/explorer-backend#350]
Currently, for every block we insert, we get its parent and check if it's part of the main chain or not and update it. We also get all blocks on the same height and put the to `main_chain = false`.

When receiving new list of blocks, we could optimize by updating all "forked" blocks on the same height at once.
 
Suppose I received blocks `C0, C1`

```
UPDATE table
SET main_chain = false
WHERE height > {C0.height}
AND chain_index = C0.chain_index
```
we  then insert `C0, C1` into DB with `main_chain = true`. We then only make the parent check for `C0`

Now I received blocks `D0, D1, D2`  (number are the same heigths as `C*`), following the same rules:
we run the above query and `C0, C1` will be set to `main_chain = false`, we then insert `D0, D1, D2` into DB with  `main_chain = true`
and update `D0`'s parent.

The optimization here is that we don't need to update main chain for the parent of `C1, D1, D2` and the blocks at the same heights.

current code is [here](https://github.com/alephium/explorer-backend/blob/42a0ab9e49c6bf2db3e271295ea4a73d780a0c30/app/src/main/scala/org/alephium/explorer/persistence/dao/BlockDao.scala#L110-L139)

**Comment by simerplaha** [comment]
Hey @tdroxler just for my understanding: the `main_chain` flag is mainly dependant on the state of the `block_hash`, but it's also maintained/set (`updateMainChainStatusAction`) in all the other 8 tables that are children of that `block_hash` (table `block_header`). 

Is this to improve read performance?

**Comment by tdroxler** [comment]
> Is this to improve read performance?

yes, there are few queries, [for example to get the balance](https://github.com/alephium/explorer-backend/blob/e73ef9ae332573d2a80d463eacf32889e8792dea/app/src/main/scala/org/alephium/explorer/persistence/queries/OutputQueries.scala#L483-L499) where we need to access only the main inputs and outputs. That would be to costly to join on the block_header to find the info.

**Comment by simerplaha** [comment]
Unassigning this for now. I'm unable to figure out how to easily make this optimisation without making sizeable changes. Will revisit once other optimisations are complete.
